shell/server/proxy.js

```diff
@@ -310,14 +321,27 @@ function openSessionInternal(grainId, user, title, apiToken) {
 
   updateLastActive(grainId, userId);
 
-  var isOwner = grainInfo.owner === userId;
+  cachedSalt = cachedSalt || Random.id(22);
+  var sessionId = generateSessionId(grainId, userId, cachedSalt);
+  var session = Sessions.findOne({_id: sessionId});
```

Comment:
The `cachedSalt` is randomly generated if not supplied. Check out `generateSessionId ` to see how it is used as part of the hash to create sessionId. There is a slight concern that if someone gained your salt, they could (maybe?) impersonate your session, but that would require a DB leak or a compromise of the client's browser.

Token revocation is the same as before. `clearSessionsAndProxies` will delete the Session from the DB, and a new Session will be established by the client which will go through the normal permissions flow.